v2/proto/voucher.go

```diff
@@ -0,0 +1,139 @@
+package proto
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+type ReservationVoucher struct {
+	// Relay is the ID of the peer providing relay service
+	Relay peer.ID
+	// Peer is the ID of the peer receiving relay service through Relay
+	Peer peer.ID
+	// Expiration is the expiration time of the reservation
+	Expiration time.Time
+	// Signature is the signature of this voucher, as produced by the Relay peer
+	Signature []byte
+}
+
+func (rv *ReservationVoucher) bytes() []byte {
+	buf := make([]byte, 1024)
+	relayBytes := []byte(rv.Relay)
+	peerBytes := []byte(rv.Peer)
```

Comment:
nit: we could avoid an allocation here if we just copied directly from the strings (which we can do).